File: engine/julia/variance_swap_copula_engine.jl
# engine/julia/variance_swap_copula_engine.jl
# OptionAlpha Agent — Module X5: Julia Volatility Derivatives & Copula Engine

module VarianceSwapCopulaEngineJulia

export calculate_realized_variance, calculate_variance_swap_greeks

function calculate_realized_variance(log_returns::Vector{Float64}, annualization_factor::Float64=252.0)
    if isempty(log_returns)
        return 0.0
    end
    sum_sq = sum(log_returns .^ 2)
    return (annualization_factor / length(log_returns)) * sum_sq
end

function calculate_variance_swap_greeks(t_years::Float64, time_elapsed::Float64, current_sigma::Float64)
    t_rem = max(1e-4, t_years - time_elapsed)
    t_safe = max(1e-4, t_years)

    cash_gamma = 2.0 / t_safe
    vega = (2.0 / t_safe) * current_sigma * t_rem
    theta = - (1.0 / t_safe) * current_sigma * current_sigma

    return (cash_gamma=cash_gamma, vega=vega, theta=theta)
end

end # module

File: engine/rust/src/variance_swap_copula_engine.rs
// engine/rust/src/variance_swap_copula_engine.rs
// OptionAlpha Agent — Module X4: Rust Volatility Derivatives & Copula Engine

pub struct VarianceSwapCopulaEngineRust;

impl VarianceSwapCopulaEngineRust {
    pub fn realized_variance(log_returns: &[f64], annualization_factor: f64) -> f64 {
        if log_returns.is_empty() { return 0.0; }
        let sum_sq: f64 = log_returns.iter().map(|&r| r * r).sum();
        (annualization_factor / log_returns.len() as f64) * sum_sq
    }

    pub fn variance_swap_greeks(t_years: f64, time_elapsed: f64, current_sigma: f64) -> (f64, f64, f64) {
        let t_rem = (t_years - time_elapsed).max(1e-4);
        let t_safe = t_years.max(1e-4);

        let cash_gamma = 2.0 / t_safe;
        let vega = (2.0 / t_safe) * current_sigma * t_rem;
        let theta = - (1.0 / t_safe) * current_sigma * current_sigma;

        (cash_gamma, vega, theta)
    }
}
